src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp

```diff
@@ -111,7 +111,10 @@ IN_PROCESS_HANDLER::NotifyDisconnect()
         m_disconnectFired = true;
     }
 
-    if (pManagedHttpContext != nullptr)
+    // Guard against IndicatemanagedRequestComplete happening before handler is invoked
+    // If m_pManagedHttpContext has changed, do not call the handler
+    // https://github.com/dotnet/aspnetcore/issues/40498
+    if (pManagedHttpContext != nullptr && pManagedHttpContext == m_pManagedHttpContext)
```

Comment:
I'm going to look at this in more detail later, but my naive thought is that you aren't going to fix a race without introducing some sort of synchronization primitive.

Reply:
Yeah we didn't want to add a new lock surrounding the old lock, this is just meant to narrow the 'window' a bit where now if  m_pManagedHttpContext changes, we prevent firing the disconnect, while before, even if m_pManagedHttpContext got completed/nulled before this if, we would still invoke it on the potentially stale pManagedHttpContext

Reply:
Why not check `m_disconnectFired` inside the lock in `IndicateManagedRequestComplete`? If it's true then we know the disconnect handler is being/will be called and the context is already stored in a local in `NotifyDisconnect`. If it's false then we'll set `m_pManagedHttpContext` to null in the lock and call `::RaiseEvent<ANCMEvents::ANCM_INPROC_MANAGED_REQUEST_COMPLETION>(m_pW3Context, nullptr)` like before, and `NotifyDisconnect` will see the null `m_pManagedHttpContext` and not call the disconnect handler.

Reply:
Would NotifyDisconnect be responsible for nulling out m_pManagedHttpContext in your proposal now? Since when m_disconnectFired  is true, if IndicateManagedRequestComplete no longer nulls out m_pManagedHttpContext , it will never become null?

Reply:
```diff
VOID
IN_PROCESS_HANDLER::IndicateManagedRequestComplete(
    VOID
)
{
    {
        SRWExclusiveLock lock(m_srwDisconnectLock);
        m_fManagedRequestComplete = TRUE;
        m_pManagedHttpContext = nullptr;
+       if (m_disconnectFired)
+       {
+           return;
+       }
    }
    ::RaiseEvent<ANCMEvents::ANCM_INPROC_MANAGED_REQUEST_COMPLETION>(m_pW3Context, nullptr);
}
```

Reply:
Doesn't this mean ANCM_INPROC_MANAGED_REQUEST_COMPLETION will never get fired?  NotifyDisconnect fires ANCM_INPROC_REQUEST_DISCONNECT

Reply:
Sure, if we care about keeping the event then do the check the other way around. Check for `m_fManagedRequestComplete` in `NotifyDisconnect` inside the lock.

Reply:
Ok, now that I've read the issue I see that the `RaiseEvent` call isn't the issue, it's the callsite of IndicateManagedRequestComplete that disposes the GCHandle and then the next allocated GCHandle can reuse the previous handle that's causing issues.

So doesn't the current change do nothing? As shown by the OPs custom logging the `pManagedHttpContext` is the same pointer across different requests.
> [2022-03-04T12:45:39.061Z, PID: 15592] [aspnetcorev2_inprocess.dll] SetManagedHttpContext ConnectionId: 11240984675822342759, pManagedHttpContext: 000002A0FABD1110
...
[2022-03-04T12:45:39.112Z, PID: 15592] [aspnetcorev2_inprocess.dll] SetManagedHttpContext ConnectionId: 4395513250809121926, pManagedHttpContext: 000002A0FABD1110

Reply:
Right pManagedHttpContext is stale, but m_pManagedHttpContext will be null for this handler right? since we don't reuse request handlers across requests and m_pManagedHttpContext will have been disposed for this situation.  This doesn't completely fix the issue since the race can happen inside of the if check with the new guard, but it should limit it a bit more

Reply:
Hi, I'm the original reporter of this bug, I know this is a bit late but I was wondering if there was a fix yet so we could remove our bodge workaround. Anyway, just for the avoidance of doubt, I agree with @BrennanConroy that this commit will not do nothing (based on my investigation as documented in the original issue #40498) .